protobuf-format.md

```diff
@@ -0,0 +1,387 @@
+
+`google.protobuf.Value` ([more info][PROTO_JSON]) is used to express a
+JSON structure. Its fields are composed of special wrapper value types
+such as `google.protobuf.BytesValue`. The protobuf runtime library
+provides JSON conversion support for these special built in message
+types.
+
+`google.protobuf.Any` ([more info][PROTO_ANY]) is used in protobuf to
+express a field whose type is not known to the enclosing message. It
+contains the binary representation of the embedded message and a URL
+string that identifies the type of the embedded message.
+
+
+#### 2.2.2 Handling instructions
+
+If the `contentType` is `application/json` or any media type with the
+structured `+json` suffix, the implementation MUST store the JSON
+payload in the `data` field.
+
+If the payload is binary, the implementation MUST store the bytes in
+the `bytes_data` field.
+
+If the payload is a protobuf, the implementation MUST store the
+payload in the `proto_data` field and the `contentType` MUST be
+`application/protobuf`.
+
+### 2.3 Extensions:
+
+`google.protobuf.Struct` ([more info][PROTO_STRUCT]) represents an
+arbitrary JSON structure. The keys of the `Struct` are the names of
+the extensions. The values of the `google.protobuf.Struct` are
+`com.protobuf.Value` ([more info][PROTO_VALUE]) fields whose contents
+correspond to the value of the extension mapped into JSON using the
+[JSON mapping rules][CE_JSON_ENCODING]. All extensions attributes MUST
+be put into this bag because they are not a part of the CloudEvents
+spec.
+
+Well known extensions can not be given a top level field because they
+have no official standing, therefore the protobuf definition can not
+commit to a strongly typed representation of the value. The protobuf
+binding must be able to map an extension name to different value
+types, e.g. a "exampleExtension" with CE type "Object" in one message
+and another "exampleExtension" with CE type "Integer" in another
+message.
+
+When a new CloudEvent spec version is released, promoted extensions
+move from the extensions bag to a strongly typed top level field.
+
+For example, consider an extension for CloudEvents 1.0:
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.0")
+  .setExtensions(
+    Struct.newBuilder()
+    .putFields(
+      "comExampleMyextension",
+      Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+In the next CloudEvents minor release, the extension is promoted to an
+official field. The protobuf message definition is updated for
+CloudEvents 1.1:
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.1")
+  .setComExampleMyextension("myvalue")
+  // During the upgrade process producers SHOULD continue setting
+  // exts bag to maintain backwards compat
+  .setExtensions(
+    Struct.newBuilder()
+      .putFields(
+        "comExampleMyextension",
+        Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+#### 2.3.1 CloudEvents upgrade process
```

Comment:
This feels out of place for this doc. If we need text like this then it should really be in the main spec since this would seem to apply in general, not just for this one serialization, right?